doc/Memory_Layout.md

```diff
@@ -1,75 +1,112 @@
-This section is written to the beginning of flash.
+The kernel code is split into two major regions. The first is `.text`, which
+holds the vector table, program code, initialization routines, and other
+read-only data.  This section is written to the beginning of flash.
+
+The second major region following up the `.text` region is the `.relocate`
+region.  It holds values that need to exist in SRAM, but have non-zero initial
+values that Tock copies from flash to SRAM as part of its initialization (see
+[Startup](Startup.md)).
 
-This is immediately followed by the `.relocate` region, which holds values the
-need to exist in SRAM, but have non-zero initial values that Tock copies from
-flash to SRAM as part of its initialization (see [Startup](Startup.md)).
+The kernel code in flash is not accessible by any process.
 
 ## Process code
 
-Processes are placed in flash starting at a known address which can be retrieved
-in the kernel using the symbol `_sapps`. Each process starts with a Tock Binary
-Format (TBF) header and then the actual application binary. Processes are placed
-continuously in flash, and each process's TBF header includes the entire size of
-the process in flash. This creates a linked-list structure that the kernel uses
-to traverse apps. The end of the valid processes are denoted by an invalid TBF
-header. Typically the flash page after the last valid process is set to all 0x00
-or 0xFF.
+Processes are placed in flash starting at a known address which can be
+retrieved in the kernel using the symbol `_sapps`. Each process starts with a
+Tock Binary Format (TBF) header and then the actual application binary.
+Processes are placed continuously in flash, and each process's TBF header
+includes the entire size of the process in flash. This creates a linked-list
+structure that the kernel uses to traverse apps. The end of the valid processes
+are denoted by an invalid TBF header. Typically the flash page after the last
+valid process is set to all 0x00 or 0xFF.
+
+Processes have read-only access to their own memory in flash, but can not read
+other processes' code. 
 
-## RAM
+# RAM
 
-RAM contains four major regions:
+The RAM holds the data currently being used by both the kernel and processes. 
+
+## Kernel RAM
+The kernel RAM contains three major regions:
 
 1. Kernel stack.
 2. Kernel data: initialized memory, copied from flash at boot.
 3. Kernel BSS: uninitialized memory, zeroed at boot.
-4. Process memory: memory space divided between all running apps.
 
+Kernel RAM is not accessible by any process.
+
+## Process RAM
+The process RAM is memory space divided between all running apps.
+
+A process's RAM contains four major regions:
+
+1. Process stack
+2. Process data
+3. Process heap
+4. Grant
+
+![Process' RAM](processram.png)
+
+A process has full access to its own stack, data and heap. However, it has no
+access to its grant region: this is only accessible by the kernel. 
```

Comment:
It may be worth clarifying that the kernel doesn't really make a stack/data/heap distinction but for debugging prints on app crashes. Furthermore the app is in total control and can use syscalls to move these around at will. Clarifying the "kernel perspective" vs "app perspective" here I think would be valuable.

Reply:
How exactly would you see clarifying this perspective? Why does the kernel not make this distinction/is it relevant?

Reply:
The kernel doesn't make this distinction because it doesn't care at all how the process organizes it's own memory. This, for example, let's a process reorder it's heap, stack, static variables if it's more efficient/simpler/or even if it doesn't have a heap at all.